ecommerce/cybersource_api.py

```diff
@@ -0,0 +1,91 @@
+"""Wrapper class for CyberSource"""
+
+import json
+import logging
+
+from ecommerce.models import Order, FulfilledOrder, Transaction
+from main.settings import ECOMMERCE_DEFAULT_PAYMENT_GATEWAY
+from ecommerce.constants import TRANSACTION_TYPE_REFUND
+from mitol.payment_gateway.api import (
+    Refund as GatewayRefund,
+    ProcessorResponse,
+    PaymentGateway,
+)
+from mitol.payment_gateway.exceptions import RefundDuplicateException
+
+log = logging.getLogger(__name__)
+
+REFUND_SUCCESS_STATES = [
+    ProcessorResponse.STATE_ACCEPTED,
+    ProcessorResponse.STATE_DUPLICATE,
+    ProcessorResponse.STATE_PENDING,
+]
+
+
+def update_order_state(order_to_update, refunded_amount, reason=None, data=None):
+    # Call order's refund method here instead
+    order_to_update.refund(refunded_amount, reason, False, data)
+    print("ORDER SUCCESSFULLY REFUNDED", str(order_to_update))
+
+
+def refund_order(order_id):
+    """
+    A function that performs refund for a given Order
+
+    Args:
+       order_id: Id of the order which is being refunded
+    """
+
+    # Get order details for the given order_id (Optimally, This should be a fulfilled order)
+    order = FulfilledOrder.objects.get(id=order_id, state=Order.STATE.FULFILLED)
+
+    # Extract transaction_id from order CyberSource json
+    order_recent_transaction = Transaction.objects.filter(order=order_id).first()
+
+    if not order_recent_transaction:
+        raise Exception(
+            f"There is no associated transaction against order_id={order_id}"
+        )
+
+    transaction_dict = json.loads(json.dumps(order_recent_transaction.data))
+
+    # This should break if there is no transaction_id in transaction_data because that's needed for refund
+    transaction_id = transaction_dict["transaction_id"]
+
+    order_amount = transaction_dict.get("req_amount")
+    order_currency = transaction_dict.get("req_currency")
+
+    refund_gateway_request = GatewayRefund(
+        transaction_id=transaction_id,
+        refund_amount=order_amount,
+        refund_currency=order_currency,
+    )
+
+    try:
+        response = PaymentGateway.start_refund(
+            ECOMMERCE_DEFAULT_PAYMENT_GATEWAY,
+            refund_gateway_request,
+        )
+        # This should be a response class from the Gateway with code and other details
+        if response.state in REFUND_SUCCESS_STATES:
+            update_order_state(
+                order,
+                response.response_data["refundAmountDetails"]["refundAmount"],
+                "",  # There is no reason/code in case of successful refund
+                response.response_data,
+            )
+
+        else:
+            print(f"THERE WAS AN ERROR WITH THE REQUEST {response.message}")
+    except RefundDuplicateException as refund_exception:
+        # Refund got an error during the API call (Check if the error was a DUPLICATE_REQUEST)
+        # We treat duplicate refund requests as a success
+        update_order_state(
```

Comment:
Since we were considering it a success, So what I had in mind regarding this was to at least create a transaction object and in this case, the API data for the transaction would not be like the success response. But your suggestion about the negative balance and the frontend inconsistency makes sense. Good call!